pandas/core/arrays/categorical.py

```diff
@@ -379,6 +382,137 @@ def __init__(
         self._dtype = self._dtype.update_dtype(dtype)
         self._codes = coerce_indexer_dtype(codes, dtype.categories)
 
+    @classmethod
+    def from_dummies(
+        cls, dummies: "DataFrame", ordered: Optional[bool] = None
+    ) -> "Categorical":
+        """Create a `Categorical` using a ``DataFrame`` of dummy variables.
+
+        The ``DataFrame`` must have no more than one truthy value per row.
+        The columns of the ``DataFrame`` become the categories of the `Categorical`.
+        A column whose header is NA will be dropped:
+        any row containing a NA value will be uncategorised.
+
+        Parameters
+        ----------
+        dummies : DataFrame
+            dtypes of columns with non-NA headers must be coercible to bool.
+            Sparse dataframes are not supported.
+        ordered : bool
+            Whether or not this Categorical is ordered.
+
+        Raises
+        ------
+        ValueError
+            If a sample belongs to >1 category
+
+        Returns
+        -------
+        Categorical
+
+        Examples
+        --------
+        >>> simple = pd.DataFrame(np.eye(3), columns=["a", "b", "c"])
+        >>> Categorical.from_dummies(simple)
+        [a, b, c]
+        Categories (3, object): [a, b, c]
+
+        >>> nan_col = pd.DataFrame(np.eye(4), columns=["a", "b", np.nan, None])
+        >>> Categorical.from_dummies(nan_col)
+        [a, b, NaN, NaN]
+        Categories (2, object): [a, b]
+
+        >>> nan_cell = pd.DataFrame(
+        ...     [[1, 0, np.nan], [0, 1, 0], [0, 0, 1]],
+        ...     columns=["a", "b", "c"],
+        ... )
+        >>> Categorical.from_dummies(nan_cell)
+        [NaN, b, c]
+        Categories (3, object): [a, b, c]
+
+        >>> multi = pd.DataFrame(
+        ...     [[1, 0, 1], [0, 1, 0], [0, 0, 1]],
+        ...     columns=["a", "b", "c"],
+        ... )
+        >>> Categorical.from_dummies(multi)
+        Traceback (most recent call last):
+            ...
+        ValueError: 1 record(s) belongs to multiple categories: [0]
+        """
+        to_drop = dummies.columns[isna(dummies.columns.values)]
+        if len(to_drop):
+            dummies = dummies.drop(columns=to_drop)
+        df = dummies.astype("boolean")
+
+        multicat_rows = df.sum(axis=1, skipna=False) > 1
+        if multicat_rows.any():
+            raise ValueError(
+                "{} record(s) belongs to multiple categories: {}".format(
+                    multicat_rows.sum(), list(df.index[multicat_rows]),
+                )
+            )
+
+        mult_by = np.arange(df.shape[1]) + 1
+        #  000            000    0   -1
+        #  010            020    2    1
+        #  001 * 1,2,3 => 003 -> 3 -> 2 = correct codes
+        #  100            100    1    0
+        codes = ((df * mult_by).sum(axis=1, skipna=False) - 1).astype("Int64")
```

Comment:
this is actually a memory heavy impl

Reply:
Sure, I guess the alternative would be something like

```python
codes = pd.Series(np.full(len(df), np.nan), dtype="Int64")

row_totals = df.sum(axis=1, skipna=False)
... # multicat_rows check

codes[row_totals == 0] = -1
row_idx, code = np.nonzero(row_totals > 0)
codes[row_idx] = code
```